runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/graph/OutputDeduplicatorTest.java

```diff
@@ -114,47 +125,66 @@ public void unchangedWithNoDuplicates() {
      * replaced with synthetic flattens.
-     * S -> A; T -> A becomes S -> A'; T -> A''; A', A'' -> Flatten -> A
+     * original graph:
+     *             --> one -> .out \
+     * red -> .out |                -> shared -> .out -> blue -> .out
+     *             --> two -> .out /
+     *
+     * fused graph:
+     *             --> [one -> .out -> shared ->] .out
+     * red -> .out |                                   (shared.out) -> blue -> .out
+     *             --> [two -> .out -> shared ->] .out
+     *
+     * deduplicated graph:
+     *             --> [one -> .out -> shared ->] .out:0 \
+     * red -> .out |                                      -> shared -> .out -> blue ->.out
+     *             --> [two -> .out -> shared ->] .out:1 /
      */
-    PTransform one =
-        PTransform.newBuilder().putInputs("in", "red.out").putOutputs("out", "one.out").build();
+    PCollection redOut = PCollection.newBuilder().setUniqueName("red.out").build();
+    PTransform red = PTransform.newBuilder().putOutputs("out", redOut.getUniqueName()).build();
     PCollection oneOut = PCollection.newBuilder().setUniqueName("one.out").build();
-    PTransform two =
-        PTransform.newBuilder().putInputs("in", "red.out").putOutputs("out", "two.out").build();
+    PTransform one =
+        PTransform.newBuilder()
+            .putInputs("in", redOut.getUniqueName())
+            .putOutputs("out", oneOut.getUniqueName())
+            .build();
     PCollection twoOut = PCollection.newBuilder().setUniqueName("two.out").build();
+    PTransform two =
+        PTransform.newBuilder()
+            .putInputs("in", redOut.getUniqueName())
+            .putOutputs("out", twoOut.getUniqueName())
+            .build();
     PCollection sharedOut = PCollection.newBuilder().setUniqueName("shared.out").build();
     PTransform shared =
         PTransform.newBuilder()
-            .putInputs("one", "one.out")
-            .putInputs("two", "two.out")
+            .putInputs("one", oneOut.getUniqueName())
+            .putInputs("two", twoOut.getUniqueName())
             .putOutputs("shared", sharedOut.getUniqueName())
             .build();
-    PTransform red = PTransform.newBuilder().putOutputs("out", "red.out").build();
-    PCollection redOut = PCollection.newBuilder().setUniqueName("red.out").build();
+    PCollection blueOut = PCollection.newBuilder().setUniqueName("blue.out").build();
     PTransform blue =
         PTransform.newBuilder()
             .putInputs("in", sharedOut.getUniqueName())
-            .putOutputs("out", "blue.out")
+            .putOutputs("out", blueOut.getUniqueName())
             .build();
-    PCollection blueOut = PCollection.newBuilder().setUniqueName("blue.out").build();
     RunnerApi.Components components =
         Components.newBuilder()
             .putTransforms("one", one)
-            .putPcollections("one.out", oneOut)
+            .putPcollections(oneOut.getUniqueName(), oneOut)
```

Comment:
Feels like there should be a one-argument overload. (Or is there any reason for the names to not agree?)

Reply:
It'd be nice, but `unique_name` is a custom field and we're just manipulating the protos directly here.